pmlc/dialect/tile/transforms/padding.cc

```diff
@@ -0,0 +1,215 @@
+};
+
+llvm::SmallVector<BoundRange, 4>
+computePaddingBounds(AffineMap access, AffineMap lower, AffineMap upper) {
+  // We only handle the 'easy' cases.
+  assert(access.getNumSymbols() == 0);
+  assert(lower.getNumSymbols() == 0);
+  assert(upper.getNumSymbols() == 0);
+  assert(lower.getNumDims() == 0);
+  assert(upper.getNumDims() == 0);
+  size_t idxs = access.getNumDims();
+  size_t size = access.getNumResults();
+  assert(lower.getNumResults() == idxs);
+  assert(upper.getNumResults() == idxs);
+  assert(access.isPureAffine());
+  std::vector<BoundRange> dimRanges;
+  for (size_t i = 0; i < idxs; i++) {
+    int64_t lowerBoundInt =
+        lower.getResult(i).cast<mlir::AffineConstantExpr>().getValue();
+    int64_t upperBoundInt =
+        upper.getResult(i).cast<mlir::AffineConstantExpr>().getValue();
+    dimRanges.emplace_back(lowerBoundInt, upperBoundInt);
+  }
+  ComputeRangeVisitor visitor(std::move(dimRanges));
+  llvm::SmallVector<BoundRange, 4> outRanges;
+  for (size_t i = 0; i < size; i++) {
+    outRanges.push_back(visitor.visit(access.getResult(i)));
+  }
+  return outRanges;
+}
+
+struct PaddingInfo {
+  llvm::SmallVector<int64_t, 4> padBelow;
+  llvm::SmallVector<int64_t, 4> padAbove;
+};
+
+static bool validForPadding(AggregationKind agg, CombinationKind comb) {
+  if (agg == AggregationKind::assign) {
+    return false;
+  }
+  if (comb == CombinationKind::none) {
+    return true;
+  }
+  if (comb == CombinationKind::mul && agg == AggregationKind::add) {
+    return true;
+  }
+  return false;
+}
+
+void PadPass::runOnFunction() {
+  auto func = getFunction();
+  llvm::DenseMap<Value, llvm::DenseMap<AggregationKind, PaddingInfo>> toPad;
+  func.walk([&](ContractionOp op) {
+    // Skip some cases where the padding pass can't operate.
+    if (op.getNumSymbols()) {
+      op.emitRemark("padding cannot be run on symbolic contractions");
+      return;
+    }
+    if (!op.lowerBounds().hasValue() || !op.upperBounds().hasValue()) {
+      op.emitRemark("contraction bounds must be computed");
+      return;
+    }
+    if (!validForPadding(op.agg(), op.combo())) {
+      op.emitRemark("invalid agg/comb for use in padding");
+      return;
+    }
+    for (size_t i = 0; i < op.getNumTensors(); i++) {
+      // Compute the largest region the source may access.
+      auto bounds = computePaddingBounds(op.getSourceMap(i), *op.lowerBounds(),
+                                         *op.upperBounds());
+      // Check if there are any out-of-bounds reads.
+      auto shape =
+          op.getTensor(i).getType().cast<mlir::MemRefType>().getShape();
+      assert(bounds.size() == shape.size());
+      bool needsPadding = false;
+      for (size_t i = 0; i < bounds.size(); i++) {
+        if (shape[i] == mlir::MemRefType::kDynamicSize) {
+          op.emitRemark("padding cannot support dynamic memref sizes");
+          return;
+        }
+        if (bounds[i].min < 0 || bounds[i].max >= shape[i]) {
```

Comment:
Drop trivial braces (I have a hard time with this one myself, trying to get into the new convention)